GitHub actions

joeapearson opened the pull request:
- [x] Matrix github actions
- [x] Add credo for linting
- [x] Mix format
- [x] Add docker compose for mongo, including replicaset support

I've added support for matrix builds.  Currently there is only one combination:

1. Ubuntu latest
2. MongoDB 4.4
2. Elixir 1.12.3
3. OTP 24.0.6

We can add more combinations later.

joeapearson committed:
Add credo

joeapearson committed:
mix format

joeapearson committed:
Add github actions

joeapearson committed:
Really basic contributing.md

joeapearson committed:
Matricise build

joeapearson committed:
Bump latest versions

joeapearson committed:
Add docker compose

joeapearson committed:
Docker compose in deamon mode

joeapearson committed:
Attempt to run tests in CI

joeapearson committed:
Use sudo

joeapearson committed:
Fix tests

joeapearson committed:
Use github provisioned mongo

joeapearson committed:
Split jobs

joeapearson committed:
Specify otp and el ver

joeapearson committed:
Set otp ver

joeapearson committed:
Run mix deps.get

joeapearson committed:
Continue on formatting and linting errors for now

joeapearson committed:
Cache build

joeapearson committed:
Cache PLTs

joeapearson committed:
Use existing bash script

joeapearson committed:
Install matrix version

joeapearson committed:
Add mongo status output

joeapearson committed:
Use start mongo script

joeapearson committed:
Split mongo version

joeapearson committed:
Allow downgrades of mongodb

joeapearson committed:
Only test Mongo 4.4 for now

joeapearson committed:
Cleanup

ankhers commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,80 @@
+on: push
```
I would suggest this being

```yaml
on:
  pull_request:
  push:
    branches:
      - master
```

This way the CI is only run for pull requests and when pushing to master.

joeapearson replied:
Alright.  Is there a reason not to want the continuous validation for every push?

ankhers commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,80 @@
+            deps
+            _build
+            priv/plts
+          key: ${{ runner.os }}-mix-${{ hashFiles('**/mix.lock') }}
```
It may be worth including the Elixir and OTP versions in the key as well. It may not be super important for code quality, seeing as you will probably only care about the most recent version of elixir.

joeapearson replied:
I think this is probably fine for code qual stuff.  If it ever becomes a problem can fix then.

ankhers commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,80 @@
+        otp-version:
+          - "24.0.6"
+        elixir-version: 
+          - "1.12.3"
```
We should probably setup an actual matrix  and test previous versions of Elixir and OTP as well.

joeapearson replied:
Yep.  Any suggestions for that?  I was starting simple so at least there is something in place as we go forward.

ankhers commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,80 @@
+          path: |
+            deps
+            _build
+          key: ${{ runner.os }}-mix-${{ hashFiles('**/mix.lock') }}
```
If there is a proper matrix of Elixir and OTP versions, you will want to include the versions in the cache key in order to pull the proper cache version.

joeapearson replied:
Yes, more important here.  I'll fix.

scottmessinger replied:
```suggestion
          key: ${{ runner.os }}-{{runner.otp-version}}-{{runner.elixir-version}}-mix-${{ hashFiles('**/mix.lock') }}
```

I think this is what @ankhers  is suggesting!

ankhers commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,80 @@
+            mongodb-org-server=${{ matrix.mongo-version.maj }}.${{ matrix.mongo-version.min }}.${{ matrix.mongo-version.patch }} \
+            mongodb-org-shell=${{ matrix.mongo-version.maj }}.${{ matrix.mongo-version.min }}.${{ matrix.mongo-version.patch }} \
+            mongodb-org-mongos=${{ matrix.mongo-version.maj }}.${{ matrix.mongo-version.min }}.${{ matrix.mongo-version.patch }} \
+            mongodb-org-tools=${{ matrix.mongo-version.maj }}.${{ matrix.mongo-version.min }}.${{ matrix.mongo-version.patch }}
```
Maybe instead of this, you can look at what [one of the official mongo drivers](https://github.com/mongodb/mongo-ruby-driver/blob/master/.github/workflows/test.yml) uses to start mongo.

joeapearson replied:
Oh excellent find.  Agreed.

ankhers commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,16 @@
+
+1. Clone this repository
+2. Install mongodb and make sure it's available on `$PATH`
+3. Run `start_mongo.bash`
```
Might be worth noting that there is also `start_mongo.bat` for windows users.

joeapearson replied:
Hmmm.. I don't see one.  Maybe there was in the past perhaps?

The contributing guide I've provided is hardly exhaustive, just meant to be a quick and dirty "get you going" for now.  In reality I've found you need to at least scan over `start_mongo.bash` and the tests to figure out what's going on.

ankhers reviewed (commented):
Overall this is fine. But there are a couple changes that would probably be beneficial.

joeapearson committed:
More selective run rules

scottmessinger committed:
Merge branch 'master' into github-actions

scottmessinger committed:
Update CI to use mongodb setup used in other drivers

scottmessinger committed:
Remove duplicate ci line

scottmessinger committed:
fix steps for ci

scottmessinger committed:
Move name

scottmessinger committed:
fix elixir version

scottmessinger committed:
set ubuntu

scottmessinger committed:
Install deps

scottmessinger committed:
Try to pass in the mongodb uri

scottmessinger committed:
Comment out text

It's unclear why this fails and it seems due to the test setup. Hopefully, we'll be able to look more into this in the future.

scottmessinger committed:
Mongo 4.2 isn't available in the setup so removing

scottmessinger committed:
Does not yet work on Mongo 5.0 and 6.0

scottmessinger committed:
Comment unused references to test we commented out

scottmessinger committed:
Change test threshold to 70

scottmessinger committed:
Make tests for 5.0 pass

scottmessinger committed:
Remove tests for putting write concern 0

It fails on Mongo 4.4 likely due to the change in Mongo 4.4 having write concern being acknowledged by default.

Future travelers who deeply care about making unacknowledged writes to the DB are welcome to look into this.

scottmessinger committed:
Remove call to deprecated function at compile time

scottmessinger committed:
Remove call to deprecated function at compile time
